Interactive selection concatenated charts e2e tests for Altair

willhuang1997 opened the pull request:
<!--
⚠️ BEFORE CONTRIBUTING PLEASE READ OUR CONTRIBUTING GUIDELINES!
https://github.com/streamlit/streamlit/wiki/Contributing
-->

## Describe your changes
- layered chart snapshots
- hconcat snapshots
- vconcat snapshots
  - can only do selection on bottom
- repeat snapshots
  - do selection non first chart
- facet snapshots
  - do selection non first chart 
  
  The playwright tests are failing but that is fixed in a separate PR: https://github.com/streamlit/streamlit/pull/8396
## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

willhuang1997 committed:
Add concatenation e2e tests

willhuang1997 committed:
Remove vega data set and add in cdn url

willhuang1997 committed:
Add a wait as the layered chart is not being rendered properly

willhuang1997 committed:
Add longer waits

willhuang1997 committed:
Fix the ordering of layering for layered chart

willhuang1997 committed:
Add snapshots

lukasmasuch commented on e2e_playwright/st_altair_chart_concatenated_select.py:
```diff
@@ -0,0 +1,155 @@
+
+chart = alt.vconcat(upper, lower)
+st.altair_chart(chart, on_select=True, key="vconcat_chart")
+if st.session_state.vconcat_chart:
```
nit: Why is this check needed? Is the `vconcat_chart` variable empty before its first selection? In that case, we might want to see if we can already fill the key with an empty datastructure as soon as the `altair_chart` command is called.

willhuang1997 replied:
There is an empty data structure but then there is an empty dataframe that already appears so we need that if statement.

lukasmasuch commented on e2e_playwright/st_altair_chart_concatenated_select_test.py:
```diff
@@ -0,0 +1,113 @@
+    # Pick Non First chart
+    themed_app.mouse.move(750, 250)
+    themed_app.mouse.down()
+    themed_app.mouse.up()
```
nit: This probably triggers a rerun or? I think it would be good to put a `wait_for_app_run(themed_app)` after this to make it more obvious. The expect below is enough since the dataframe only appears after the first selection, but that might be something we want to change.

lukasmasuch replied:
The same might also apply to some of the other tests below.

lukasmasuch reviewed (approved):
LGTM 👍 A couple of comments above

willhuang1997 committed:
Only apply widget id if on_select is true (#8396)

<!--
⚠️ BEFORE CONTRIBUTING PLEASE READ OUR CONTRIBUTING GUIDELINES!
https://github.com/streamlit/streamlit/wiki/Contributing
-->

## Describe your changes
- title
- this mainly fixes the e2e tests that are failing because there are
stale elements that are showing up
- technically, if https://github.com/streamlit/streamlit/pull/8400 is
merged, this PR is not needed
## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

willhuang1997 committed:
Fix nits

willhuang1997 committed:
Fix nits

willhuang1997 committed:
Merge branch 'interactiveChartsAltair/e2e' into interactiveChartsAltair/e2e_concatenated

willhuang1997 committed:
Merge branch 'interactiveChartsAltair/e2e' into interactiveChartsAltair/e2e_concatenated
